Спринт 15 Добавление Clean Architecture

KateDesu opened the pull request.

Commit:
Спринт 15 Добавление Clean Architecture

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/Creator.kt:
```diff
@@ -0,0 +1,17 @@
+import com.practicum.playlistmaker.domain.api.TracksRepository
+import com.practicum.playlistmaker.domain.impl.TracksInteractorImpl
+
+class Creator {
```
Рекомендация:
Стоит поменять `class ` на `object` так, что бы объект `Creator` имел единственный инстанс на всё приложение.

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,30 @@
+        .addConverterFactory(GsonConverterFactory.create())
+        .build()
+
+    private val iTunesService = retrofit.create(ITunesSearchApi::class.java)
```
Рекомендация:
Это зависимость, которую стоит создавать в `Creator` и передавать сюда через конструктор класса.

Commit:
Спринт 15 Внесение небольших изменений

Commit:
Спринт 15 Переключение на LF

Commit:
Спринт 15 Добавление работы с SearchHistory по типу чистой архитектуры

Commit:
Спринт 15 Добавление лямбда-выражения в TracksAdapter

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/search/SearchActivity.kt:
```diff
@@ -0,0 +1,257 @@
+    private lateinit var clearHistoryButton: Button
+    private lateinit var flProgressBar: FrameLayout
+
+    private lateinit var tracksInteractor: TracksInteractor
```
Рекомендация:
Будет лучше инициализировать это поле через `by lazy`, тогда не получится обратиться к этому полю до инициализации и как следствие приложения не упадёт с ошибкой `NullPointerException`
Получится:
```
private val tracksInteractor: TracksInteractor by lazy {
    Creator.provideTracksInteractor()
}
```
Для `searchHistoryInteractor` **аналогично**

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/tracks/TracksAdapter.kt:
```diff
@@ -1,15 +1,22 @@
-package com.practicum.playlistmaker
+package com.practicum.playlistmaker.ui.tracks
```
Рекомендация:
`TracksAdapter` и `TracksViewHolder` всё же относится к экрану поиска, а не к экрану плеера. Оба стоит перенести в `com.practicum.playlistmaker.ui.search`

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/tracks/TracksAdapter.kt:
```diff
@@ -1,15 +1,22 @@
 import com.google.gson.Gson
+import com.practicum.playlistmaker.PLAYLISTMAKER_PREFERENCES
+import com.practicum.playlistmaker.R
+import com.practicum.playlistmaker.data.dto.SearchHistory
```
Рекомендация:
Стоит удалить лишние импорты

IliaAnK commented on app/src/main/java/com/practicum/playlistmaker/ui/settings/SettingsActivity.kt:
```diff
@@ -11,6 +11,9 @@ import androidx.core.view.ViewCompat
 import androidx.core.view.WindowInsetsCompat
 import com.google.android.material.appbar.MaterialToolbar
 import com.google.android.material.switchmaterial.SwitchMaterial
+import com.practicum.playlistmaker.App
```
Обязательно:
Для признака темы, так же как и для остальных фич, нужно сформировать свои `domain` и `data` слои.
Тогда его запись и чтение в `SettingsActivity` и `App` будут происходить через интерактор.

`SettingsActivity` не должна обращаться для этого к `App`, всё взаимодействие только через общую кодовую базу `domain` и `data` слоёв

Commit:
Спринт 15 Удаление зависимости iTunesService и исправление некоторых ошибок

Commit:
Спринт 15 Приведение SharedPreferences к чистой архитектуре

Commit:
Спринт 15 Исправление бага пустой истории в ui

Commit:
Спринт 15 Передача в адаптер списка треков
Исправление багов: обновление списка треков после его добавления, отображение ошибки потери интернета

Commit:
Спринт 15 Чистка кода и переключение на LF
